pkg/volume/flexvolume/plugin.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flexvolume
+
+import (
+	"path"
+	"strings"
+	"sync"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/mount"
+	utilstrings "k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+const flexVolumePluginName = "kubernetes.io/flexvolume"
+
+// FlexVolumePlugin object.
+type flexVolumePlugin struct {
+	driverName string
+	execPath   string
+	host       volume.VolumeHost
+	runner     exec.Interface
+
+	sync.Mutex
+	unsupportedCommands []string
+}
+
+var _ volume.AttachableVolumePlugin = &flexVolumePlugin{}
+var _ volume.PersistentVolumePlugin = &flexVolumePlugin{}
+
+// Init is part of the volume.VolumePlugin interface.
+func (plugin *flexVolumePlugin) Init(host volume.VolumeHost) error {
+	plugin.host = host
+	// call the init script
+	call := plugin.NewDriverCall(initCmd)
+	_, err := call.Run()
+	return err
+}
+
+func (plugin *flexVolumePlugin) getExecutable() string {
+	parts := strings.Split(plugin.driverName, "/")
+	execName := parts[len(parts)-1]
+	return path.Join(plugin.execPath, execName)
+}
+
+// Name is part of the volume.VolumePlugin interface.
+func (plugin *flexVolumePlugin) GetPluginName() string {
+	return plugin.driverName
+}
+
+// GetVolumeName is part of the volume.VolumePlugin interface.
+func (plugin *flexVolumePlugin) GetVolumeName(spec *volume.Spec) (string, error) {
+	call := plugin.NewDriverCall(getVolumeNameCmd)
+	call.AppendSpec(spec, plugin.host, nil)
+
+	status, err := call.Run()
+	if isCmdNotSupportedErr(err) {
+		return (*pluginDefaults)(plugin).GetVolumeName(spec)
+	} else if err != nil {
+		return "", err
+	}
+	return utilstrings.EscapeQualifiedNameForDisk(status.Device), nil
```

Comment:
Is `status.Device` the best field to get a response for a `getVolumeNameCmd`? In `examples/volumes/flexvolume/README.md` (which should be updated along with this or soon afterwards), it says: `device` is the `<Path to the device attached. This field is valid only for attach calls>`